src/aoc2020/day1.clj

```diff
@@ -1,61 +1,102 @@
 (ns aoc2020.day1
-  (:require [clojure.java.io :as io]
-            [clojure.string :as str]))
-
-(defn file->seq
-  [filename]
-  (-> filename
-      (io/resource)
-      (slurp)
-      (str/split-lines)))
-;;Q1: find 2 numbers in the file that would sum up to be 2020, and find the product of these 2 numbers.
-
-;Approach 1: brute force
-;loop through the entries and use 2020 to minus each number, and loop through the rest of the vectors to find the
-;if the (2020 - num) exist. Time: O(n^2) space: O(1)
-
-
-;Approach 2: binary search
-;sort the vector in ascending order, and use 2 pointers pointing to head and end of the vector, and move pointers
-;based on the sum of 2 numbers. Time: O(log n) space: O(1)
-
-(defn binary-search
-  [data-set]
-  (->> map (fn [left-pointer
-                right-pointer]
-             ((if (== 2020 (+ (get data-set left-pointer)
-                              (get data-set right-pointer)))
-
-                )))
-       (range 0 (-> data-set count))
-       (range (-> data-set count) 0 -1))
-  )
+  (:require [aoc2020.util :refer [file->seq]]))
+
+(defn complement
+  [sum x]
+  (- sum x))
+
+
+(defn find-complement
+  [data-set sum]
+  (let [compl-exist?  (fn [num]
+                        (contains? data-set (complement sum num)))
+        result        (filterv compl-exist? data-set)]
+
+    (when (first result)
+      result)))
+
+(defn two-sum-product
+  [data-s sum]
+  (let [data-set     (into #{} data-s)
+        result       (find-complement data-set sum)
+        num1         (result 0)
+        num2         (result 1)]
+    (* num1 num2)))
+
+
+(defn compl-product
+  [data-s sum num]
+  (let [data-set     (into #{} data-s)
+        compl        (complement sum num)
+        result       (find-complement data-set compl)
+        num2         (get result 0)
+        num3         (get result 1)]
+
+    (when (and num2 num3)
+      (* num num2 num3))))
+
+
+(defn three-sum-product
+  [data-s sum]
+  (->> (mapv
+         (fn [x] (compl-product data-s sum x))
+         data-s)
+       (filterv int?)
+       (first)))
 
 
 (comment
-  (def entries (file->seq "aoc2020/day1/input.txt"))
-  #_=>
-  ["1348"
-   "1621"
-   "1500"
-   "1818"
-   "1266"
-   ...]
-
-  (def sorted-entries (->> entries
-                           (map (fn [string] (Integer/parseInt string)))
-                           (sort)))
-  #_=>
-  (183
-    455
-    470
-    651
-    667
-    695
-    934
-    974
-    988
-    ...)
+  ;;------------------------function evaluation---------------------------
+
+  (complement 2020 20)
+  #_=> 2000
+  (mapv (partial complement 2020) [1 6 5 23 90 3])
+  #_=> [2019 2014 2015 1997 1930 2017]
+  (mapv (partial complement 2020) sample-entry-s)
+  #_=> [299 1041 1654 1721 1345 564]
```

Comment:
These feel like copy-paste examples.   i.e., the third is not particularly useful to demonstrate the function, while the second is much more useful and self-contained. The first example is just a duplicate subset of the second... why is it still there? Is 2020 the best sum to illustrate this example? Are `[1 6 5 23 90 3]` the examples you would given to demonstrate how something works, as well as edge cases?

Same idea applies below